src/puny.cpp

```diff
@@ -0,0 +1,385 @@
+      if (iterator == input.end()) {
+        return false;
+      }
+      char code_point = *iterator++;
+      int digit = ((code_point <= 'z') && (code_point >= 'a'))
+                      ? code_point - 'a'
+                      : (((code_point <= '0') && (code_point >= '0'))
+                             ? code_point - '0' + 26
+                             : -1);
+      if (digit < 0) {
+        return false;
+      }
+      i = i + digit * w;
+      int t = k <= bias ? 1 : k >= bias + 26 ? 26 : k - bias;
+      if (digit < t) {
+        break;
+      }
+      w = w * (36 - t);
+    }
+    bias = adapt(i - start_i, count + 1, start_i == 0);
+    n = n + i / (count + 1);
+    i = i % (count + 1);
+    if ((n < 0x80) || is_disallowed(n)) {
+      return false;
+    }
+
+    count++;
+    i++;
+  }
+
+  return true;
+}
+
+// We return the empty string on error.
+std::optional<std::string> convert_domain_to_puny(std::string_view input, bool be_strict) {
+  std::string out;
+  (void)be_strict; // currently ignored.
+  // We seem to be allowing an infinite domain size and an infinite number of
+  // of labels. This is not correct. A domain is limited to 255 characters
+  // and labels cannot exceed 64 characters. However, wpt_tests seems to want
+  // to allow more general cases?
+  // Though wpt_tests does not want limits, let us put one anyhow. If someone
+  // has a domain with over 1MB, we refuse to work on it (safety!).
+  if(input.size() > 1'000'000) { return std::nullopt; }
+  while (!input.empty()) {
+    size_t loc_dot = input.find('.');
+    size_t loc_full_stop = input.find("\u3002"); // complete as needed
+    size_t loc = std::min(loc_dot, loc_full_stop);
+    size_t utf8_sep_size = loc == loc_dot ? 1 : 3;
+    bool is_last_label = (loc == std::string_view::npos);
+    size_t label_size = is_last_label ? input.size() : loc;
+    size_t label_size_with_dot =
+        is_last_label ? input.size() : loc + utf8_sep_size;
+    std::string_view label_view =
+        input.substr(0, label_size); // does not contain the dot.
+    if(label_view.empty()) {
+        // nothing to do.
+    // next check is not as expensive as it looks:
+    } else if (ada::checkers::begins_with(label_view, "xn--") ||
+        ada::checkers::begins_with(label_view, "XN--") ||
+        ada::checkers::begins_with(label_view, "Xn--") ||
+        ada::checkers::begins_with(label_view, "xN--")) {
+      // It is already punycode, we just validate it!
+      if (!verify_punycode(label_view)) {
+        return std::nullopt;
+      }
+      out += label_view;
+    } else {
+      // It is not punycode, we have to do the hard work.
+      // This could be made faster by not creating a new temporary string.
+      std::optional<std::string> label_string =
+          utf8_to_punycode(label_view.data(), label_view.size());
+
+      if (!label_string.has_value()) {
+        return std::nullopt;
+      }
+      if (label_string.value().empty()) {
```

Comment:
this can be simplified to label_string.value_or.empty(). Removes the previous branch